packages/contracts/contracts/voting/multisig/Multisig.sol

```diff
@@ -0,0 +1,432 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.10;
+
+import {ERC165Upgradeable} from "@openzeppelin/contracts-upgradeable/utils/introspection/ERC165Upgradeable.sol";
+import {Initializable} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import {CountersUpgradeable} from "@openzeppelin/contracts-upgradeable/utils/CountersUpgradeable.sol";
+
+import {PluginUUPSUpgradeable} from "../../core/plugin/PluginUUPSUpgradeable.sol";
+import {IDAO} from "../../core/IDAO.sol";
+import {_uncheckedAdd, _uncheckedSub} from "../../utils/UncheckedMath.sol";
+import {IMajorityVoting} from "../majority/IMajorityVoting.sol";
+import {Addresslist} from "../addresslist/Addresslist.sol";
+
+/// @title Multisig
+/// @author Aragon Association - 2022.
+/// @notice The on-chain multisig governance plugin in which a proposal passes if X out of Y approvals are met.
+/// @dev This contract inherits from `MajorityVotingBase` and implements the `IMajorityVoting` interface.
+contract Multisig is Initializable, ERC165Upgradeable, Addresslist, PluginUUPSUpgradeable {
+    using CountersUpgradeable for CountersUpgradeable.Counter;
+
+    /// @notice A container for proposal-related information.
+    /// @param executed Wheter the proposal is executed or not.
+    /// @param parameters The proposal-specific approve settings at the time of the proposal creation.
+    /// @param tally The approve tally of the proposal.
+    /// @param approvers The approves casted by the approvers.
+    /// @param actions The actions to be executed when the proposal passes.
+    struct Proposal {
+        bool open;
+        bool executed;
+        ProposalParameters parameters;
+        Tally tally;
+        mapping(address => bool) approvers;
+        IDAO.Action[] actions;
+    }
+
+    /// @notice A container for the proposal parameters.
+    /// @param minApprovals The number of approvals required.
+    /// @param snapshotBlock The number of the block prior to the proposal creation.
+    struct ProposalParameters {
+        uint256 minApprovals;
```

Comment:
The `addresslistLength` that uses OZs `Checkpoints` lib return values as `uint256`.
We could do it but do more casting and overflow checks. Do you want to add it to the remaining task https://aragonassociation.atlassian.net/browse/APP-1417 ?

Reply:
Would be good. Even if OZ uses uint256 and we use uint64, it would automatically do the casting. And the arguments for minapprovals in the function would be uint64. So no need to do checks manually

Reply:
@brickpop If you agree, please add it to the description of https://aragonassociation.atlassian.net/browse/APP-1417

Reply:
@Michael-A-Heuer that's a purely technical discussion, my engineering opinion here is second to yours, guys.